src/ray/gcs/gcs_server/ray_syncer.h

```diff
@@ -0,0 +1,129 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <memory>
+#include <type_traits>
+
+#include "ray/common/asio/periodical_runner.h"
+#include "ray/gcs/gcs_server/gcs_resource_report_poller.h"
+#include "ray/gcs/gcs_server/grpc_based_resource_broadcaster.h"
+
+namespace ray {
+namespace syncer {
+
+// RaySyncer is a service to sync components in the cluster.
+// It's supposed to be used to synchronize resource usage and scheduling information
+// in ray cluster. This component is still in developing.
+class RaySyncer {
+ public:
+  RaySyncer(instrumented_io_context &main_thread,
+            std::unique_ptr<::ray::gcs::GrpcBasedResourceBroadcaster> braodcaster,
+            std::unique_ptr<::ray::gcs::GcsResourceReportPoller> poller)
+      : ticker_(main_thread),
+        broadcaster_(std::move(braodcaster)),
+        poller_(std::move(poller)) {}
+
+  void Start() {
+    poller_->Start();
+    broadcast_thread_ = std::make_unique<std::thread>([this]() {
+      SetThreadName("resource_report_broadcaster");
+      boost::asio::io_service::work work(broadcast_service_);
+      broadcast_service_.run();
+    });
+    ticker_.RunFnPeriodically(
+        [this] {
+          auto beg = resources_buffer_.begin();
+          auto ptr = beg;
+          static auto max_batch = RayConfig::instance().resource_broadcast_batch_size();
+          for (size_t cnt = resources_buffer_proto_.batch().size();
+               cnt < max_batch && cnt < resources_buffer_.size(); ++ptr, ++cnt) {
+            resources_buffer_proto_.add_batch()->mutable_data()->Swap(&ptr->second);
+          }
+          resources_buffer_.erase(beg, ptr);
+          broadcast_service_.dispatch(
+              [this, resources = std::move(resources_buffer_proto_)]() mutable {
+                broadcaster_->SendBroadcast(std::move(resources));
+              },
+              "SendBroadcast");
+          resources_buffer_proto_.Clear();
+        },
+        RayConfig::instance().raylet_report_resources_period_milliseconds(),
+        "RaySyncer.deadline_timer.report_resource_report");
+  }
+
+  void Stop() {
+    poller_->Stop();
+    if (broadcast_thread_ != nullptr) {
+      broadcast_service_.stop();
+      if (broadcast_thread_->joinable()) {
+        broadcast_thread_->join();
+      }
+    }
+  }
+
+  // External API
+  template <typename T>
+  void Update(T update) {
```

Comment:
should update `NodeResourceChange` and `ResourcesData` belong to separate method? looks they are doing different things.

i think one nice thing of documenting the APIs: if we can not easily describe the APIs by simple comments, probably we should separate them

Reply:
@scv119  this is not the final one we'll have. I'll comment it there.